Compiling with `2.24.0` fails: `module not found: org.jspecify`

## Description

I have a modular (Jigsaw) project that uses log4j2. Compilation fails after updating to 2.24.0 from 2.23.1 with the error "".

## Configuration

**Version:** 2.24.0

**Operating system:** macOS 14.6.1

**JDK:** 

openjdk 21.0.4 2024-07-16 LTS
OpenJDK Runtime Environment (build 21.0.4+9-LTS)
OpenJDK 64-Bit Server VM (build 21.0.4+9-LTS, mixed mode, sharing)


## Logs

Running my gradle build, I get this:
```
23:21:38: Executing 'build'...

> Task :utility-logging:processResources NO-SOURCE
> Task :utility-fx-web:generateCatalogAsToml FROM-CACHE
> Task :generateCatalogAsToml FROM-CACHE
> Task :utility-db:generateCatalogAsToml FROM-CACHE
> Task :utility-logging:generateCatalogAsToml FROM-CACHE
> Task :utility:generateCatalogAsToml FROM-CACHE
> Task :utility-fx-icons-ikonli:generateCatalogAsToml FROM-CACHE
> Task :utility-fx-db:generateCatalogAsToml FROM-CACHE
> Task :utility-fx:generateCatalogAsToml FROM-CACHE
> Task :utility-fx-web:processResources NO-SOURCE
> Task :utility-fx-controls:generateCatalogAsToml FROM-CACHE
> Task :compileJava NO-SOURCE
> Task :utility-fx-db:processResources NO-SOURCE
> Task :utility-db:processResources
> Task :utility-fx:processResources
> Task :utility-fx-controls:processResources
> Task :utility-fx-icons-ikonli:processResources

> Task :utility-fx-icons:compileJava FAILED

> Task :utility-logging:sourcesJar
> Task :utility-fx-icons-ikonli:sourcesJar
> Task :utility-fx-db:sourcesJar
> Task :utility-fx-web:sourcesJar
> Task :utility-fx:sourcesJar
> Task :utility-fx-controls:sourcesJar
error: module not found: org.jspecify
1 error
> Task :utility-db:sourcesJar
> Task :cabe

> Task :utility:compileJava FAILED
23 actionable tasks: 14 executed, 9 from cache
error: module not found: org.jspecify
1 error

FAILURE: Build completed with 2 failures.

1: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':utility-fx-icons:compileJava'.
> Compilation failed; see the compiler error output for details.

* Try:
> Run with --info option to get more log output.
> Run with --scan to get full insights.
==============================================================================

2: Task failed with an exception.
-----------
* What went wrong:
Execution failed for task ':utility:compileJava'.
> Compilation failed; see the compiler error output for details.

* Try:
> Run with --info option to get more log output.
> Run with --scan to get full insights.
==============================================================================

BUILD FAILED in 448ms
23:21:39: Execution finished 'build'.
```

## Reproduction

Sorry, still have not had the time to create a minimal project. I think any project using modules working with 2.23.1 should expose the error  - I still need to investigate. But changing the log4j2 version is the only change I did before running into this error and changing back to 2.23.1 fixed it. Compile with version 2.23.1, then try again using 2.24.0.

My project does (obviously) not use jspecify, adding it as a dependency will probably fix it.

I see this in the log4j-core module-info.class and think jspecify should be a transitive dependency and the static might have to be removed:
```
        requires static transitive org.jspecify;
```

--- a/log4j-api/pom.xml
+++ b/log4j-api/pom.xml
@@ -54,7 +54,9 @@
       <!-- Used in StringBuilders through reflection -->
       java.sql;static=true,
       <!-- Used in ProcessIdUtil through reflection -->
-      java.management;static=true
+      java.management;static=true,
+      <!-- Remove `transitive` for optional dependencies -->
+      org.jspecify;transitive=false
     </bnd-extra-module-options>
 
   </properties>

--- a/log4j-core/pom.xml
+++ b/log4j-core/pom.xml
@@ -98,6 +98,7 @@
       java.naming;transitive=false,
       org.apache.commons.csv;transitive=false,
       org.fusesource.jansi;transitive=false,
+      org.jspecify;transitive=false,
       org.zeromq.jeromq;transitive=false,
       <!-- A module descriptor is only available in version 1.2.16+, hence it is not detected -->
       com.conversantmedia.disruptor;substitute="disruptor";transitive=false;static=true,

--- a/log4j-parent/pom.xml
+++ b/log4j-parent/pom.xml
@@ -53,6 +53,7 @@
     <asciidoctor-maven-plugin.version>2.2.4</asciidoctor-maven-plugin.version>
     <docker-maven-plugin.version>0.45.0</docker-maven-plugin.version>
     <exam-maven-plugin.version>4.13.5</exam-maven-plugin.version>
+    <gmavenplus-plugin.version>3.0.2</gmavenplus-plugin.version>
     <maven-taglib-plugin.version>2.4</maven-taglib-plugin.version>
     <!-- `surefire.version` property used in `apache.org:apache`: -->
     <surefire.version>3.5.0</surefire.version>
@@ -990,6 +991,68 @@
                   </excludes>
                 </bannedDependencies>
               </rules>
+            </configuration>
+          </execution>
+        </executions>
+      </plugin>
+
+      <plugin>
+        <groupId>org.codehaus.gmavenplus</groupId>
+        <artifactId>gmavenplus-plugin</artifactId>
+        <version>${gmavenplus-plugin.version}</version>
+        <dependencies>
+          <dependency>
+            <groupId>org.codehaus.groovy</groupId>
+            <artifactId>groovy-ant</artifactId>
+            <version>${groovy.version}</version>
+            <scope>runtime</scope>
+          </dependency>
+          <dependency>
+            <groupId>org.codehaus.groovy</groupId>
+            <artifactId>groovy</artifactId>
+            <version>${groovy.version}</version>
+            <scope>runtime</scope>
+          </dependency>
+        </dependencies>
+        <executions>
+          <execution>
+            <id>ban-static-transitive</id>
+            <goals>
+              <goal>execute</goal>
+            </goals>
+            <phase>verify</phase>
+            <configuration>
+              <continueExecuting>false</continueExecuting>
+              <scripts>
+                <script><![CDATA[
+                  import java.io.StringWriter
+                  import java.util.spi.ToolProvider
+
+                  if ("jar" != project.packaging) {
+                      log.info("Skipping module descriptor check, since the project type is not `jar`.")
+                      return
+                  }
+                  String jarFile = project.build.directory + "/" + project.build.finalName + ".jar";
+                  if (!new File(jarFile).exists()) {
+                      log.info("Skipping module descriptor check, since `" + jarFile + "` is missing.")
+                      return
+                  }
+                  StringWriter out = new StringWriter()
+                  StringWriter err = new StringWriter()
+                  ToolProvider jar = ToolProvider.findFirst("jar").orElseThrow()
+                  int result = jar.run(new PrintWriter(out), new PrintWriter(err), "-d", "-f", jarFile)
+                  if (result != 0) {
+                      throw new RuntimeException("Failed to decompile the module descriptor in `" + jarFile + "`:\n" + err)
+                  }
+                  log.debug("Module descriptor: " + out)
+                  for (String line : out.toString().split("\r?\n", -1)) {
+                      if (line.contains("static") && line.contains("transitive")) {
+                          throw new RuntimeException("The `static` and `transitive` modifiers should not be use together: " + line)
+                      }
+                  }
+                  log.info("Successfully verified module descriptor in `" + jarFile + "`.")
+              ]]></script>
+              </scripts>
             </configuration>
           </execution>
         </executions>

--- a/log4j-to-jul/pom.xml
+++ b/log4j-to-jul/pom.xml
@@ -34,6 +34,10 @@
       <!-- Annotations only -->
       org.jspecify.*;resolution:=optional
     </bnd-extra-package-options>
+    <bnd-extra-module-options>
+      <!-- Remove `transitive` for optional dependencies -->
+      org.jspecify;transitive=false
+    </bnd-extra-module-options>
   </properties>
 
   <dependencies>

--- a/log4j-to-slf4j/pom.xml
+++ b/log4j-to-slf4j/pom.xml
@@ -45,6 +45,10 @@
       <!-- This bridge also support SLF4J 2.x -->
       org.slf4j.*;version="${slf4j.support.range}"
     </bnd-extra-package-options>
+    <bnd-extra-module-options>
+      <!-- Remove `transitive` for optional dependencies -->
+      org.jspecify;transitive=false
+    </bnd-extra-module-options>
 
     <slf4j2.version>2.0.16</slf4j2.version>
   </properties>

